runtime/src/iree-amd-aie/driver/hsa/command_buffer.cc

```diff
@@ -0,0 +1,413 @@
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_hsa_native_executable_entry_point_kernel_info(
+              executable, entry_point, &kernel_info));
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_resource_set_insert(command_buffer->resource_set, 1,
+                                       &executable));
+
+  // Creating a packet to store the command
+  hsa::hsa_amd_aie_ert_packet_t* cmd_pkt = nullptr;
+  IREE_HSA_RETURN_IF_ERROR(
+      command_buffer->hsa_symbols,
+      hsa_amd_memory_pool_allocate(device_allocator->global_kernarg_mem_pool,
+                                   PACKET_SIZE, /*flags*/ 0,
+                                   reinterpret_cast<void**>(&cmd_pkt)),
+      "hsa_amd_memory_pool_allocate");
+  cmd_pkt->state = hsa::HSA_AMD_AIE_ERT_STATE_NEW;
+  // # of arguments to put in command
+  // there's an extra leading word or something like that...
+  cmd_pkt->count = 1 + 5 + (2 * bindings.count);
+  cmd_pkt->opcode = hsa::HSA_AMD_AIE_ERT_START_CU;
+  cmd_pkt->header.AmdFormat = hsa::HSA_AMD_PACKET_TYPE_AIE_ERT;
+  cmd_pkt->header.header = hsa::HSA_PACKET_TYPE_VENDOR_SPECIFIC
+                           << hsa::HSA_PACKET_HEADER_TYPE;
+
+  // Creating the payload for the packet
+  hsa::hsa_amd_aie_ert_start_kernel_data_t* cmd_payload = nullptr;
+  uint32_t cmd_handle;
+
+  IREE_HSA_RETURN_IF_ERROR(command_buffer->hsa_symbols,
+                           hsa_amd_get_handle_from_vaddr(
+                               reinterpret_cast<void*>(cmd_pkt), &cmd_handle),
+                           "hsa_amd_get_handle_from_vaddr");
+  IREE_HSA_RETURN_IF_ERROR(
+      command_buffer->hsa_symbols,
+      hsa_amd_memory_pool_allocate(device_allocator->global_kernarg_mem_pool,
+                                   PACKET_SIZE, /*flags*/ 0,
+                                   reinterpret_cast<void**>(&cmd_payload)),
+      "hsa_amd_memory_pool_allocate");
+
+  // Selecting the PDI to use with this command
+  cmd_payload->cu_mask = 1;
+  // Transaction opcode
+  cmd_payload->data[0] = 3;
+  // unused or ?
+  cmd_payload->data[1] = 0;
+  cmd_payload->data[2] = kernel_info.ipu_inst_handle;
+  // unused or ?
+  cmd_payload->data[3] = 0;
+  cmd_payload->data[4] = kernel_info.num_instr;
+
+  for (iree_host_size_t j = 0; j < bindings.count; ++j) {
+    hsa_device_pointer_t device_buffer = iree_hal_hsa_buffer_device_pointer(
+        iree_hal_buffer_allocated_buffer(bindings.values[j].buffer));
+    uint32_t handle;
+    IREE_HSA_RETURN_IF_ERROR(
+        command_buffer->hsa_symbols,
+        hsa_amd_get_handle_from_vaddr(device_buffer, &handle),
+        "hsa_amd_get_handle_from_vaddr");
+    cmd_payload->data[5 + (2 * j)] = handle;
+    cmd_payload->data[5 + (2 * j) + 1] = 0;
+  }
+
+  cmd_pkt->payload_data = reinterpret_cast<uint64_t>(cmd_payload);
+
+  // TODO(max): this doesn't go here, should go in pending_actions maybe?
+  // like shouldn't be done right at dispatch time
+  // Configure the queue's hardware context.
```

Comment:
Don't you need the context before submitting anything into the queue? I'm not familiar with pending actions. 

One thing I noticed is that, when using the python bindings, if you configured the dispatch early on, and then ran the compilation, there was a bug that would kill the context in the device upon execution of the subprocess (From python). This is because somehow upon the subcommand finishing, it would call an operation that would force a flush in the file descriptor of the driver, and that would kill the context. 

So early creation of the context had some implications in that case. But this is really a corner case.